internal/lido/contracts/csfeedistributor/rewards.go

```diff
@@ -0,0 +1,197 @@
+This function is responsible for:
+retrieving non-claimed rewards for Lido CSM node
+params :-
+network (string): The name of the network (e.g."holesky").
+nodeID (*big.Int): Node Operator ID
+proofStrings ([]string): Merkle Proof of Node Operator Rewards
+returns :-
+a. *big.Int
+Non-claimed rewards
+b. error
+Error if any
+*/
+func Rewards(network string, nodeID *big.Int, proofStrings []string) (*big.Int, error) {
+	var rewards *big.Int
+	contract, err := csFeeDistributorContract(network)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call csFeeDistributorContract: %w", err)
+	}
+
+	treeCID, err := treeCID(network)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call treeCID: %w", err)
+	}
+
+	shares, err := cumulativeFeeShares(treeCID, nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call cumulativeFeeShares: %w", err)
+	}
+
+	proof, err := convertToByte(proofStrings)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call convToByte: %w", err)
+	}
+
+	fees, err := contract.GetFeesToDistribute(nil, nodeID, shares, proof)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call GetFeesToDistribute: %w", err)
+	}
+
+	bondInfo, err := bond.BondSummary(network, nodeID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call BondSummary: %w", err)
+	}
+
+	rewards = new(big.Int).Add(bondInfo.Excess, fees)
+	return rewards, nil
+}
+
+func cumulativeFeeShares(treeCID string, nodeID *big.Int) (*big.Int, error) {
+	// Get tree data from IPFS
+	treeData, err := treeData(treeCID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to call treeData: %v", err)
+	}
+
+	// Compare nodeOperatorID in tree with nodeId to get shares
+	// Binary search for the nodeOperatorId
+	low, high := 0, len(treeData.Values)-1
+	for low <= high {
+		mid := (low + high) / 2
+		nodeOperatorId, err := convertToBigInt(treeData.Values[mid].Value[0])
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert nodeOperatorId: %v", err)
+		}
+		cmp := nodeOperatorId.Cmp(nodeID)
+		if cmp == 0 {
+			// Node operator ID matches, return the shares
+			shares, err := convertToBigInt(treeData.Values[mid].Value[1])
+			if err != nil {
+				return nil, fmt.Errorf("failed to convert shares: %v", err)
+			}
+			return shares, nil
+		} else if cmp < 0 {
+			low = mid + 1
+		} else {
+			high = mid - 1
+		}
+	}
+	return nil, fmt.Errorf("invalid nodeId")
+}
+
+func treeCID(network string) (string, error) {
+	var treeCIDString string
+	contract, err := csFeeDistributorContract(network)
+	if err != nil {
+		return treeCIDString, fmt.Errorf("failed to call csFeeDistributorContract: %w", err)
+	}
+
+	treeCIDString, err = contract.TreeCid(nil)
+	if err != nil {
+		return treeCIDString, fmt.Errorf("failed to call TreeCid: %w", err)
+	}
+	return treeCIDString, nil
+}
+
+func treeData(treeCID string) (Tree, error) {
+	var treeData Tree
+	gatewayURL := fmt.Sprintf("https://ipfs.io/ipfs/%s", treeCID) // Public gateway URL
+	resp, err := http.Get(gatewayURL)
```

Comment:
Use our [`utils.GetRequest`](https://github.com/NethermindEth/sedge/blob/develop/internal/utils/request.go#L42) to make the request using [exponential backoff retries](https://en.wikipedia.org/wiki/Exponential_backoff)

Reply:
[Done](https://github.com/NethermindEth/sedge/pull/401/commits/27f7757898e1b136611fd13fcdd4d5bc8fcb39ce)